_posts/2024-08-27-cheri-myths-safe-languages.markdown

```diff
@@ -0,0 +1,156 @@
+When you're writing code with a set of non-malicious collaborators, this is a big benefit.
+
+But what happens with the rest of the supply chain?
+The [Rudra paper](https://dl.acm.org/doi/10.1145/3477132.3483570) used some static analysis to find 264 memory-safety bugs in cargo crates, and issued 76 CVEs and 112 RustSec advisories as a result.
+These were difficult to spot unless you knew the exact incorrect idiom to look for, so a cursory glance at `unsafe` code in your dependencies would not have helped.
+
+This is the best case.
+The bugs that Rudra found were accidental.
+The [cve-rs project](https://github.com/Speykious/cve-rs/tree/main?tab=readme-ov-file) shows how to exploit a soundness issue in the Rust borrow checker to implement any kind of memory-safety bug entirely without `unsafe`.
+Even a type-safe language may have implementation bugs that make it possible to sneak in malicious code.
+
+This is not just a problem for Rust.
+Go is a garbage-collected type-safe language, but it exposes a slice type that contains a base, length, and capacity.
+This is not atomically updated and so you can have two goroutines race storing to a shared variable that is a slice type and a third will eventually see the base of one and the length of the other, allowing out-of-bounds accesses, which can then be used to escape type safety.
+The Java VM is infamous for security vulnerabilities that allow programs to break type safety in various ways.
+
+The more complex a type system is, whether it's enforced statically or dynamically, the more likely that there will be a bug that allows an attacker to exploit it.
+When you're pulling in a million-line dependency, auditing it to make sure that it doesn't trigger such a vulnerability is impossible.
+
+Java shipped with a lot of complex infrastructure for sandboxing components.
+The [`SecurityManager`](https://docs.oracle.com/javase/8/docs/api/java/lang/SecurityManager.html) let you restrict the privileges of running code and elevate privileges later.
+This depended on the correctness of most of the rest of the JVM.
+In Android, Google removed it (technically it's still there, you just got an exception if you installed a non-default policy) and made the process boundary the only security boundary.
+They did this because anything that depends on the correctness of the entire JVM for security is likely to be broken.
+The same rationale should apply to compartmentalisation models that rely on other languages' type-safety guarantees.
+
+Using CHERI to isolate parts of the supply chain does not have this problem.
+The security properties of a CHERI ISA are simple and local.
+They have been formally verified at the ISA level on two CHERI variants and the CHERIoT Ibex core has had its implementation formally verified.
+This is possible because the properties are simple (slightly oversimplifying):
```

Comment:
I don't like the use of "simple" again, but I do not have a suggestion.

Reply:
"because the properties can be expressed simply as:"